From 61b2d43ed9b562e55b3eaf9ab9efa44f97b91b86 Mon Sep 17 00:00:00 2001
From: Andrew Pinski <apinski@marvell.com>
Date: Mon, 7 Oct 2019 17:53:05 -0700
Subject: [PATCH 368/767] arm64: Add workaround for Marvell erratum 37119

commit a8735dfd13887dac412adf0d8be69904486c0a60 from
git@git.assembla.com:cavium/WindRiver.linux.git

Enable workaround for erratum 37119.  On some OcteonTX2 chips,
while using the SSO workslots in userspace can cause problems
as the cache are not invalidated correctly.  To workaround
this erratum, a local tlbi is placed before the eret.
This TLBI does not need to invalidate anything so a tlbi against
asid 0 is used.

Change-Id: I0313daae25e8ccaaffb84f9b6db5e6859b693dfa
Signed-off-by: Andrew Pinski <apinski@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/17086
Reviewed-by: Stanislaw Kardach <Stanislaw.Kardach@cavium.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[Kevin: Use ERRATA_MIDR_RANGE_LIST macro to fix the duplicate entry warning
for capability.]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 arch/arm64/Kconfig               | 13 +++++++++++++
 arch/arm64/include/asm/cpucaps.h |  3 ++-
 arch/arm64/kernel/cpu_errata.c   | 17 +++++++++++++++++
 arch/arm64/kernel/entry.S        |  3 +++
 4 files changed, 35 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index 153fdf69d3f5..0c2270ebcf30 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -609,6 +609,19 @@ config CAVIUM_ERRATUM_36890
 
 	  If unsure, say Y.
 
+config MRVL_ERRATUM_37119
+	bool "Marvell erratum 37119"
+	default y
+	help
+	  Enable workaround for erratum 37119.  On some OcteonTX2 chips,
+	  while using the SSO workslots in userspace can cause problems
+	  as the cache are not invalidated correctly.  To workaround
+	  this erratum, a local tlbi is placed before the eret.
+	  This TLBI does not need to invalidate anything so a tlbi against
+	  asid 0 is used.
+
+	  If unsure, say Y.
+
 config QCOM_FALKOR_ERRATUM_1003
 	bool "Falkor E1003: Incorrect translation due to ASID change"
 	default y
diff --git a/arch/arm64/include/asm/cpucaps.h b/arch/arm64/include/asm/cpucaps.h
index f78f309da11f..782e5403027c 100644
--- a/arch/arm64/include/asm/cpucaps.h
+++ b/arch/arm64/include/asm/cpucaps.h
@@ -53,7 +53,8 @@
 #define ARM64_HAS_DCPODP			43
 #define ARM64_WORKAROUND_1463225		44
 #define ARM64_WORKAROUND_CAVIUM_36890		45
+#define ARM64_WORKAROUND_MRVL_37119		46
 
-#define ARM64_NCAPS				46
+#define ARM64_NCAPS				47
 
 #endif /* __ASM_CPUCAPS_H */
diff --git a/arch/arm64/kernel/cpu_errata.c b/arch/arm64/kernel/cpu_errata.c
index 1c99122c413e..6959c0e59e66 100644
--- a/arch/arm64/kernel/cpu_errata.c
+++ b/arch/arm64/kernel/cpu_errata.c
@@ -691,6 +691,16 @@ static const struct midr_range cavium_erratum_36890_cpus[] = {
 };
 #endif
 
+#ifdef CONFIG_MRVL_ERRATUM_37119
+static const struct midr_range mrvl_erratum_37119_cpus[] = {
+	/* Marvell OcteonTX 2, 96xx pass A0, A1, and B0 */
+	MIDR_RANGE(MIDR_MRVL_OCTEONTX2_96XX, 0, 0, 1, 0),
+	/* Marvell OcteonTX 2, 95 pass A0/A1 */
+	MIDR_RANGE(MIDR_MRVL_OCTEONTX2_95XX, 0, 0, 0, 1),
+	{},
+};
+#endif
+
 #ifdef CONFIG_QCOM_FALKOR_ERRATUM_1003
 static const struct arm64_cpu_capabilities qcom_erratum_1003_list[] = {
 	{
@@ -809,6 +819,13 @@ const struct arm64_cpu_capabilities arm64_errata[] = {
 		ERRATA_MIDR_RANGE_LIST(cavium_erratum_36890_cpus),
 		.cpu_enable = cpu_enable_trap_zva_access,
 	},
+#endif
+#ifdef CONFIG_MRVL_ERRATUM_37119
+	{
+		.desc = "Marvell erratum 37119",
+		.capability = ARM64_WORKAROUND_MRVL_37119,
+		ERRATA_MIDR_RANGE_LIST(mrvl_erratum_37119_cpus),
+	},
 #endif
 	{
 		.desc = "Mismatched cache type (CTR_EL0)",
diff --git a/arch/arm64/kernel/entry.S b/arch/arm64/kernel/entry.S
index 84a822748c84..4db2bf6a0e84 100644
--- a/arch/arm64/kernel/entry.S
+++ b/arch/arm64/kernel/entry.S
@@ -364,6 +364,9 @@ alternative_else_nop_endif
 	ldr	lr, [sp, #S_LR]
 	add	sp, sp, #S_FRAME_SIZE		// restore sp
 
+#ifdef CONFIG_MRVL_ERRATUM_37119
+alternative_insn "tlbi aside1, xzr", "nop", ARM64_WORKAROUND_MRVL_37119
+#endif
 	.if	\el == 0
 alternative_insn eret, nop, ARM64_UNMAP_KERNEL_AT_EL0
 #ifdef CONFIG_UNMAP_KERNEL_AT_EL0
-- 
2.31.1

